File: Drone/scripts/upForward.py
#!/usr/bin/env python2



import rospy, time
#from mavros_msgs.msg import Pose
from geometry_msgs.msg import PoseStamped , TwistStamped
from mavros_msgs.srv import SetMode, CommandBool
from mavros_msgs.msg import AttitudeTarget, State
from MavrosSetUP import MavrosSetUP
import std_msgs.msg
from tf.transformations import quaternion_from_euler
from tf.transformations import euler_from_quaternion




def main():

    setPoints(0,0,2)
    mavrosSet.set_mode("OFFBOARD", 5)  # sets drone to offboard which is custom (0)
    mavrosSet.set_arm(True, 5)  # arms the drone

    while not rospy.is_shutdown():

        odomSubscriber()
        #move(0,0,.2)
        print("PoseZ: ", poseZ)
        if(poseZ >= 5):
            move(0,0,0)
            print("Hi")
        rate.sleep()


setPointMsg = PoseStamped()
def setPoints(x,y,z):
    for x in range(100 and not rospy.is_shutdown()):
        setPointMsg.pose.position.x = x
        setPointMsg.pose.position.y = y
        setPointMsg.pose.position.z = z
        setPointMsg.header.seq = 0
        setPointMsg.header.stamp = rospy.rostime.Time.now()
        setPointMsg.header.frame_id = '' #iris::base_link

        setPointPub.publish(setPointMsg)

        rate.sleep()

moveMsg = TwistStamped()
def move(x, y, z): #set speed
    moveMsg.twist.linear.x = x
    moveMsg.twist.linear.y = y
    moveMsg.twist.linear.z = z
    #moveMsg.thrust = 1
    #print("Publishing")
    movePub.publish(moveMsg)

def odomSubscriber():
    rospy.Subscriber('/mavros/local_position/pose', PoseStamped, poseCall)

def poseCall(data):
    global poseX
    global poseZ

    poseX = data.pose.position.x
    poseZ = data.pose.position.z


if __name__== "__main__":
    rospy.init_node('moveupForward', anonymous=True)
    movePub = rospy.Publisher('/mavros/setpoint_attitude/cmd_vel', TwistStamped, queue_size=10)
    setPointPub = rospy.Publisher('/mavros/setpoint_position/local', PoseStamped, queue_size=10)
    mavrosSet = MavrosSetUP()
    mavrosSet.setUp()

    rate = rospy.Rate(30)
    poseX = 0.0
    poseZ = 0.0


    main()

File: Drone/scripts/upForward.cpp
#include <ros/ros.h>
#include <geometry_msgs/PoseStamped.h>
#include <mavros_msgs/CommandBool.h>
#include <mavros_msgs/State.h>
#include <mavros_msgs/SetMode.h>


mavros_msgs::State currentState;
void stateCall(const mavros_msgs::State::ConstPtr& msg) {
	currentState = *msg; //stores all state data in currentState
}
int main(int argc, char **argv) {
	ros::init(argc, argv,"setupDrone");
	ros::NodeHandle nh;
	ros::Rate rate(30.0);

	ros::Subscriber stateSub = nh.subscribe<mavros_msgs::State>("/mavros/state", 10, stateCall);
	ros::Publisher posPub = nh.advertise<geometry_msgs::PoseStamped>("/mavros/setpoint_position/local", 10);
	ros::ServiceClient armSrv = nh.serviceClient<mavros_msgs::CommandBool>("/mavros/cmd/arming");
	ros::ServiceClient modeSrv = nh.serviceClient<mavros_msgs::SetMode>("mavros/set_mode");



	while(ros::ok() && !currentState.connected) {
		ros::spinOnce();
		rate.sleep();
	}

	geometry_msgs::PoseStamped sendPose;
	sendPose.pose.position.x = 0;
	sendPose.pose.position.y = 0;
	sendPose.pose.position.z = 2;

	for(int i = 0; ros::ok() && i < 100; i++) {
		posPub.publish(sendPose);
		ros::spinOnce();
		rate.sleep();
	}

	mavros_msgs::SetMode setMode;
	setMode.request.custom_mode = "OFFBOARD"; //Creates a SetMode srv-msg with offboard string

	mavros_msgs::CommandBool setArm;
	setArm.request.value = true;

	ros::Time lastRequest = ros::Time::now();

	while(ros::ok()) {
		if(currentState.mode != "OFFBOARD" //if mode hasnt been set yet and its been longer than X seconds
				&& (ros::Time::now() - lastRequest > ros::Duration(1.0))) {
			if(modeSrv.call(setMode) && setMode.response.mode_sent) { //send the mode and if received
				ROS_INFO("Offboard enabled");
			}
			lastRequest = ros::Time::now();
		}
		else {
			if(!currentState.armed &&
					(ros::Time::now() - lastRequest) > ros::Duration(1.0)) {
				if(armSrv.call(setArm) && setArm.response.success) {
					ROS_INFO("Drone Armed");
				}
				lastRequest = ros::Time::now();
			}

		}
		ros::spinOnce();
		rate.sleep();
	}

	return 0;
}

